### PR TITLE
Adds support for clear cache

### LINKED ISSUE #1454: Add developer option to clear cache
Requested by @ScottWitte - 
"As a developer, I need to be certain that the changes I make in code and upload are indeed being used when viewing in FxR. If a change doesn't work, is it because I didn't really fix it or the old code is stuck in cache?

BTW, I really hate Safari wiping out everything when I just want a clean cache. Please don't do THAT.

### PR DESCRIPTION
**Pull request by keianhzo**
Fixes #1454 Adds support for clearing the cache. At the moment just clears all the caches until with have a more complete spec to do it more granular. Also I've refactored sessions code and fixed:
- Not all sessions were being recreated for methods that required session recreation
- Removed the intermediate session stack as it's not really necessary we can handle it with the linked hashmap

### CODE REVIEW

File: app/src/common/shared/org/mozilla/vrbrowser/browser/engine/SessionStack.java [hunk 1]
```diff
@@ -803,8 +815,29 @@ protected void setTrackingProtection(final boolean aEnabled) {
...
+            // https://mozilla.github.io/geckoview/javadoc/mozilla-central/org/mozilla/geckoview/StorageController.html#clearData-long-
+            for (Map.Entry<Integer, SessionState> entry : mSessions.entrySet()) {
+                SessionState state = entry.getValue();
+                if (state != null) {
+                    state.mSession.stop();
+                    state.mSession.close();
+                }
             }
+
+            mRuntime.getStorageController().clearData(StorageController.ClearFlags.ALL).then(aVoid -> {
```
Comment (bluemarvin) [inline app/src/common/shared/org/mozilla/vrbrowser/browser/engine/SessionStack.java]: This does more than just clear the cache. I think `StorageController.ALL_CACHES` would be better? If we want to purge all data I think it should be called something other than clear cache?

Comment (keianhzo) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/engine/SessionStack.java]: Moved to Privacy and updated to ALL_CACHES.

**Comment by keianhzo** [comment]
@MortimerGoro @bluemarvin I've updated with the change request and also made some refactoring to remove the intermediate session stack that is not really necessary and that simplifies the sessionstack code a bit.

**Comment by bluemarvin** [comment]
@keianhzo we should probably wait to land this until the UI spec has been updated? @jvonitter

**Comment by keianhzo** [comment]
@bluemarvin There was never UI spec for this one, was just adding a clear cache button. You mean to have a more complete spec make the "clear" more granular?

**Comment by bluemarvin** [comment]
> @bluemarvin There was never UI spec for this one, was just adding a clear cache button. You mean to have a more complete spec make the "clear" more granular?

Right, I was asking if we should wait for the spec with more details or land this and follow up once there is a spec.

**Comment by jvonitter** [comment]
yes please wait for the spec. it's on Nadja's list and it'll be done soon

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -290,6 +290,17 @@
     <!-- The string labels an On/Off switch in the developer options dialog and is used to toggle enabling Servo. -->
     <string name="developer_options_servo">Enable Servo</string>
 
+    <!-- The string labels the description text for the clear cookies and site data button in the
+         privacy options dialog. -->
+    <string name="developer_options_clear_cache_cookies_site_description">Cookies &amp; SiteData</string>
```
Comment (bluemarvin) [inline app/src/main/res/values/strings.xml]: Should `SiteData` be `Site Data`?

Comment (keianhzo) [reply app/src/main/res/values/strings.xml]: Fixed

**Comment by keianhzo** [comment]
We have to leave out the description from the spec as GV doesn't currently expose any API for getting the cache used disk space.

**Comment by ScottWitte** [comment]
I'm the original requester for this feature and I'm a little concerned. Am I right that you will have one button that clears cache, history, and cookies, much as with Mobile Safari? If so, in my experience that would be a mistake. A major reason for clearing the cache is so you can reload a page and be certain you have the most current content. That is especially important when developing and is just about the first thing any support or FAQ tells a user to do when troubleshooting. When something doesn't work as expected you can be certain it isn't just because some old code is still being used. 

If you nuke everything just to clear the cache you force users to retype URLs, which is tedious and error-prone, especially in a VR interface, or to search for lost URLs, which can be difficult. I suggest you use the approach of all other browsers (except mobile Safari) and give users control over just what they want to clear. Look at mobile FireFox as a model.

Thanks for your consideration. Really excited to see how FxR is developing!

**Comment by bluemarvin** [comment]
There are two buttons, one clears cache, one clears cookies and data.

**Comment by ScottWitte** [comment]
Excellent. Thanks for the clarification.

**Pull request merged**